### ISSUE TITLE
Filtering on many to many

### ISSUE THREAD

**Issue by RamiAwar** [sqlalchemy-filterset/sqlalchemy-filterset#11]
Hello! First issue here! 

I was wondering how I can use this library to filter on many to many relationships, didn't find an example in the docs. Is it even possible to do this with the current API? Or do I need to define a custom join strategy that takes in the through table model and join clause perhaps?

**Comment by ymezencev** [comment]
**The library is not intended to completely replace sqlalchemy, but rather to simplify standard queries.**

The correct approach would be to figure out how to do this in pure sql, and then try to simplify the query using the library. 
MtM filtering can be unobvious and should be considered separately in each case.


Example:
We have: company table, city table and company_city mtm table
Write query that: Get all cities of company by company_id/company_name

1) Denormalise data to one array field and filter by it.
2) Using subquery IN
```
class CityFilterSet(BaseFilterSet):  
    id = Filter(city_table.c.id)  
    name = Filter(city_table.c.name)  
    company_id = MethodFilter(method="filter_by_company_id")  
    company_name = MethodFilter(method="filter_by_company_name")  
  
    @staticmethod  
    def filter_by_company_id(query: Select, value: uuid.UUID | None) -> Select:  
        if value:  
            subquery = select(company_city_table.c.city_id).where(  
                company_city_table.c.company_id == value  
            )  
            return query.where(city_table.c.id.in_(subquery))  
        return query  
  
    @staticmethod  
    def filter_by_company_name(query: Select, value: uuid.UUID | None) -> Select:  
        if value:  
            subquery = (  
                select(company_city_table.c.city_id)  
                .join(company_table, company_table.c.id == company_city_table.c.company_id)  
                .where(company_table.c.name == value)  
            )  
            return query.where(city_table.c.id.in_(subquery))  
        return query
```

```
query = select(City)
filter_set = CityFilterSet(query)
query = filter_set.filter_query({"company_id": value})
```

```
select * 
from city 
where city_id in (select city_id from company_city where company_id == :value);
```
We can not filter by several fields of subquery in different filters (company_id + company_name) because this will generate multiple subqueries, which is inefficient.
This approach could be used if there are not so many rows in the subquery.

3) Use two queries and join/(subquery IN) them
```
select *  
from city  
join (select distinct company_city.city_id  
from company_city  
join company on company_city.company_id = company.id  
where company.name = 'some_company_name'  -- many company filters here
) city_subquery on city_subquery.city_id = city.id;
```

```
select *  
from city  
where city_id in (select distinct company_city.city_id  
from company_city  
join company on company_city.company_id = company.id  
where company.name = 'some_company_name' -- many company filters here
)
```

We can do this with two filtersets CompanyFilterSet and CityFilterSet. Build two different queries and then join them by selecting distinct ids in many to many. 

```
city_query = ...  # build query using CityFilterSet
  
if company_params:  
    company_query = ...  # build query using CopmanyFilterSet
    company_city_query = select(distinct(company_city.c.id)).join(  
        company_query, company_query.c.id == company_city.c.company_id  
    )  
    city_query = city_query.join(  
        company_city_query, company_city_query.c.city_id == city_query.c.id  
    )
```

**Comment by ymezencev** [comment]
Thanks for suggestion with MultiJoinStrategy, let's take a closer look at how the filtering will be done

When we have the query:
```
SELECT project
```
we receive a list of projects.

However, when we use the query:
```
SELECT project
JOIN company_city ON company_city.project_id = project.id
JOIN city ON company_city.city_id = city.id
WHERE city.count_people > 10
```
we no longer get a list of projects, but rather a list of company_city.

Filtering via a many-to-many relationship through a join will require grouping or using DISTINCT. Therefore, filtering in the way described is not directly feasible.

**Comment by ymezencev** [comment]
`Or even better, adding a submodule that provides Pydantic filtersets that plug into FastAPI and Starlette directly?`


How do you see this refinement, what do you see missing in the current functionality?

Now to make everything work, we need to make:
1) one dto to describe what queries parameters we expect QueryParamsSchemaDto
2) one dto to describe the filtering scheme FilterSchemaDto
3) Declare a filterset
4) Write a query in the database and put a filterset on it

The code is layered and universal for any frameworks
Here is an example:
https://sqlalchemy-filterset.github.io/sqlalchemy-filterset/examples/fastapi/


In fact, you can make a generic repository and specify basic queries (get, list, create, update) in it using filtersets
But such things are specific for each team, so we didn't put it into the general functionality. Perhaps we should make an example of such implementation and add it to the documentation.

**Comment by RamiAwar** [comment]
Yeah indeed your approach is more general.

Here is what I did. I still think a lot of the work I did can be implemented here, as I build on top of the sqlalchem-filterset design.

But I chose to have a single Pydantic model be the filterset + input schema (to skip the extra step you have here).
https://github.com/RamiAwar/pydantic-sqlalchemy-filterset

I'm not sure it's a great idea to integrate the pydantic part, but the work I did on joins can be nice to add. Also, more examples are useful (many to manys and such).

**Comment by Mrreadiness** [comment]
Hi @RamiAwar!
Could you please give us more details about `MultiJoinStrategy`. 
How do you handle deduplication?
Do you use `distinct` or sqlaclchemy [`unique()`](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.unique) or maybe something else?

**Comment by RamiAwar** [comment]
Hey @Mrreadiness ! I don't handle it in the filtersets actually. I think this is the user's responsibility.

For me the filterset must only modify the 'query' (sqlalchemy select or whatever).
The query execution should be handled by the users of the filtersets. So they're responsible to call .unique() in all places if their filterset includes joins at all.

**Comment by Mrreadiness** [comment]
@RamiAwar I've found, that we can chain `RelationJoinStrategy`, [check it out](https://github.com/sqlalchemy-filterset/sqlalchemy-filterset/pull/15/files#diff-85594ee1c08d5ca836d2cf8566fc73e25de841f139d0d00ce39babd0ce16aed4R44). 
I also discovered some join-deduplication issues, but it's not connected to many-to-many logic.

This solution is limited by inner joins (as RelationJoinStrategy). From a "regular point of view", I don't see how outer join can be used, feels counterintuitive. 

I guess we can also implement something with a similar interface for variants 2 and 3 from [this comment](https://github.com/sqlalchemy-filterset/sqlalchemy-filterset/issues/11#issuecomment-2016952859).

But I'm still doubting what is worth adding and what is not. Many-to-many filtering is a complicated thing, and every time I face it - ended up with a specific query-tuning solution. 
@RamiAwar Do you have some common patterns of many to many filtering in your project?
@ymezencev Your thoughts?

**Comment by RamiAwar** [comment]
My 2 cents: Idk about the technical feasibility of what I'm about to say, but it shouldn't be the barrier.

I think we should keep the api as close to SQLAlchemy as possible. The less we deviate, the more this library will likely be the best out there. No one wants to learn another api and DSL, they will prefer building their own makeshift solution and use sqlalchemy which they know and love.

What does this mean practically?
API should support .outerjoin(...) and .innerjoin(...), with an optional onclause and chaining (and other stuff sqlalchemy supports if we want but this is the minimum).

**Comment by Mrreadiness** [comment]
> I think we should keep the api as close to SQLAlchemy as possible

Yeah, I agree. 
Maybe something like this would be closer to the target:
```python
class JoinChainStrategy(BaseStrategy, ABC):
    def __init__(
        self,
        *chain: RelationJoinStrategy,
    ) -> None:
        self.chain = chain

    def filter(self, query: Select, expression: Any) -> Select:
        for el in self.chain:
            query = el.apply_join(query)
        return query.where(expression)
```

About outerjoin-s. We had been thinking about it and had not found usage for it in the context of filters. Anyway, adding them is not a big deal, just inherit  `RelationJoinStrategy` and override `_build_join`. (If you have examples with outer joins from real live - I'm ready to consider adding to lib)

Required onclause made on purpose. I think it is a good example of `Explicit is better than implicit`, plus it's easier to handle the deduplication of joins. 

The goal in general is not to wrap the sqlalchemy with support of all functionality, it's about simplification of building filters (a.k. most common scenarios of filters)

**Comment by RamiAwar** [comment]
Yeah I actually use outer joins in my work all the time. I'll try to find an example, but they are definitely needed. Sometimes with optional filters you just gotta outer join, inner would filter out results you want.

Why do you care much about the join type though? Did you see what I did here? 
https://github.com/RamiAwar/pydantic-sqlalchemy-filterset/blob/main/pydantic_sqlalchemy_filterset/strategies.py

**Comment by ymezencev** [comment]
In the context of the filterset library, the primary goal of the JoinRelationStrategy is to add joins only when a filter value is provided, keeping the query simple and efficient when no filters are applied:

```
-- Simple query without unnecessary joins
SELECT A.* 
FROM A;
```

When a filter value is provided, the necessary join is dynamically added to fetch the filtered results efficiently:

```
-- Adding a join only when a filter value is specified
SELECT A.* 
FROM A
JOIN B ON A.b_id = B.id
WHERE B.some_value = 'desired_value';
```

This strategy avoids performance degradation by not including joins that aren't needed. Including unnecessary joins can complicate and slow down the query.

Using an outer join, such as a LEFT JOIN, when planning to filter out null values can be counterintuitive and inefficient. Since we filter out rows where the join condition results in null anyway, it's simpler and more efficient to use a regular join:

```
-- Regular join to filter out rows with null values
SELECT A.* 
FROM A
JOIN B ON A.b_id = B.id
WHERE B.some_value = 'desired_value';
```

If the requirement is to find rows in A that do not have corresponding entries in B, we should use a query specifically designed for that purpose:

```
-- Finding rows in A without corresponding entries in B
SELECT A.* 
FROM A
WHERE A.b_id IS NULL;
```

In summary, an outer join does not properly align with the concept of a join that can be optionally applied for filtering purposes.

**Comment by ymezencev** [comment]
Oh, there is a valid scenario for using a LEFT JOIN

```
SELECT A.* 
FROM A
LEFT JOIN B ON A.b_id = B.id
WHERE B.id IS NULL OR B.some_value = 'desired_value';
```
Query retrieves all rows from table A that either:
1. Do not have a corresponding entry in table B (i.e., B.id IS NULL)
2. Have a connection to table B where B.some_value is equal to 'desired_value'.

**Comment by RamiAwar** [comment]
I can give you a generic example that's very common:

Let's say you have Blog and Tags.
Many to many relationship, so you'll create a jump table: BlogTag.
Relationship is optional however. Some blogs don't have any tags!

Sometimes you're searching across multiple fields (ex. Blog search) and you want to search blog titles + tags. That's a perfect use case for a filterset.

`http://myblog.com?q=fastapi&category=technical`

Here q searches across title + tags, and since tags are optional we need to join them with an outer join when filtering.
Category on the other hand is straightforward, let's say we have a Category model and we just use an inner join on it (1-1).

**Comment by Mrreadiness** [comment]
@RamiAwar You convinced me. With this improvement `RelationOuterJoinStrategy` will also be added.
[Current prototype](https://github.com/sqlalchemy-filterset/sqlalchemy-filterset/pull/15).
However, I faced this [bug](https://github.com/sqlalchemy-filterset/sqlalchemy-filterset/issues/16). I don't think that we should add JoinChainStrategy before fixing it, because the fix may change API of `Strategis` (BTW open to discussion).

**Comment by RamiAwar** [comment]
That's why I didn't implement join dedupe 😛 Didn't want to deal with this stuff..

Side comment: I don't love the naming, it's complicated. 
`JoinChainStrategy` and `RelationOuterJoinStrategy` and ...

Would be much more readable to have:
`MultiJoinStrategy` and `OuterJoinStrategy`

Users don't need to know anything about the fact that we're chaining, this is a dev detail for us.
'Relation' is kind of implicit, you're never joining on 'non-relations'.

**Comment by Mrreadiness** [comment]
Hi @RamiAwar!
I guess [pr](https://github.com/sqlalchemy-filterset/sqlalchemy-filterset/pull/15) on the final stage. What do you think? Does it meet your needs?

**Comment by Mrreadiness** [comment]
MultiJoinStrategy was added at 2.3.0. 
@RamiAwar Thank you for your help!

**Comment by RamiAwar** [comment]
Nice work! @Mrreadiness 🙂
